### PR TITLE
Endian fixes

### PR DESCRIPTION
**Pull request by Schlumpf**
As @willytax mentioned in #174, there are some problems on big-endian systems. I also use SOEM on a big-endian PowerPC system and found a few bugs. No guarantee, that there aren't more.

### CODE REVIEW

File: soem/ethercatconfig.c [hunk 1]
```diff
@@ -482,8 +483,8 @@ int ecx_config_init(ecx_contextt *context, uint8 usetable)
             context->slavelist[slave].SM[1].StartAddr = htoes(context->slavelist[slave].mbx_ro);
             context->slavelist[slave].SM[1].SMlength = htoes(context->slavelist[slave].mbx_rl);
             context->slavelist[slave].SM[1].SMflags = htoel(EC_DEFAULTMBXSM1);
-            context->slavelist[slave].mbx_proto =
-               ecx_readeeprom2(context, slave, EC_TIMEOUTEEP);
+            val = ecx_readeeprom2(context, slave, EC_TIMEOUTEEP);
```
Comment (nakarlsson) [inline soem/ethercatconfig.c]: skip adding uint32_t val, use the same look'n'feel as done earlier in the file, for example L388 "eedat = etohl(ecx_readeeprom2(context, slave, EC_TIMEOUTEEP)); /* read mailbox offset */"

File: soem/ethercattype.h [hunk 2]
```diff
@@ -526,6 +525,8 @@ typedef struct
   #define etohll(A) (A)
 
 #elif !defined(EC_LITTLE_ENDIAN) && defined(EC_BIG_ENDIAN)
+  #define put_unaligned32(val, ptr)        \
+    (memcpy((ptr), ((uint32*)&(val))+1, 4))
```
Comment (nakarlsson) [inline soem/ethercattype.h]: Please explain how put_unaligned32 works for BE.

Why don't you need put put_unaligned64 fro BE

File: soem/ethercatconfig.c (hunk 1 above)
Comment (Schlumpf) [reply soem/ethercatconfig.c]: I used the temporary variable because `etohs()` is a macro. Fully expanded, it results in (on big-endian systems):
`((((uint32)(ecx_readeeprom2(context, slave, 20000)) & 0xff000000) >> 24) | \ (((uint32)(ecx_readeeprom2(context, slave, 20000)) & 0x00ff0000) >> 8) | \ (((uint32)(ecx_readeeprom2(context, slave, 20000)) & 0x0000ff00) << 8) | \ (((uint32)(ecx_readeeprom2(context, slave, 20000)) & 0x000000ff) << 24))`.
So `ecx_readeeprom2()` would called 4 times, instead of once.

Comment (nakarlsson) [reply soem/ethercatconfig.c]: Ok, that's no good, that would need to be addressed.  A few of those and other ethercat commands but isolated to ethercatconfig.c, would you mind fixing it in this PR?

File: soem/ethercattype.h (hunk 2 above)
Comment (Schlumpf) [reply soem/ethercattype.h]: If I understood correctly, `put_unaligned32()` should copy the used 4 bytes of an 8 byte buffer variable to a destination buffer. It is used in `ecx_siigetbyte()`, only.

If you write an 32 bit value to an 64 bit variable, the highest 4 bytes are used:
Value: `0x12345678` -> Memory: `78 56 34 12 00 00 00 00`
On big-endian systems, value use the last 4 bytes:
Value: `0x12345678` -> Memory: `00 00 00 00 12 34 56 78`
If you only copy the first 4 bytes of the variable on big-endian systems, you would get 0x0 every time. Here you have to copy the last 4 bytes.

This change differentiates between the systems and copies the second 4 bytes on big-endian systems. I cannot use the  `put_unaligned64()` macro, because it copies 8 bytes instead of 4.

Comment (nakarlsson) [reply soem/ethercattype.h]: The reason for using put_alignedX is that &(context->esibuf[eadr << 1]) might not be aligned, that might trigger exceptions on MCU arch that can't handle that.

So, if I understand you correct, the issue is only put_aligned32 due to the fact that edat is an uint64_t I don't think changing put_unaligned32 is the right way to go, would this work?

            put_unaligned32(edat, &(context->esibuf[eadr << 1]));
            edat = htoell(edat);
            cnt = 4;

Comment (Schlumpf) [reply soem/ethercattype.h]: That wouldn't be correct. `htoell()` would swap all bytes, so the order of the needed 4 bytes would changed.
The problem is the use of a `uint64` as a buffer for a 32 bit value. In my new commit, I solved the problem otherwise by shifting the needed bits to the upper 4 bytes.

File: soem/ethercatconfig.c (hunk 1 above)
Comment (Schlumpf) [reply soem/ethercatconfig.c]: Done in new commit

File: soem/ethercattype.h [hunk 3]
```diff
@@ -513,10 +513,10 @@ typedef struct
 #define put_unaligned64(val, ptr)        \
   (memcpy((ptr), &(val), 8))
 
-#if !defined(EC_BIG_ENDIAN) && defined(EC_LITTLE_ENDIAN)
-  #define put_unaligned32(val, ptr)        \
-    (memcpy((ptr), &(val), 4))
+#define put_unaligned32(val, ptr)        \
```
Comment (nakarlsson) [inline soem/ethercattype.h]: Can you revert the changes in this file all together? Doesn't need to be part of the PR anymore.?

File: soem/ethercatmain.c [hunk 4]
```diff
@@ -370,6 +370,10 @@ uint8 ecx_siigetbyte(ecx_contextt *context, uint16 slave, uint16 address)
          /* 4 byte response */
          else
          {
+#ifdef EC_BIG_ENDIAN
```
Comment (nakarlsson) [inline soem/ethercatmain.c]: I get you point but I'd like to make it more clear what's going on. How about making it more explicit?

   uint64 edat64;
   uint32 edat32;
...
         edat64 = ecx_readeepromFP (context, configadr, eadr, EC_TIMEOUTEEP);
         /* 8 byte response */
         if (context->slavelist[slave].eep_8byte)
         {
            put_unaligned64(edat64, &(context->esibuf[eadr << 1]));
            cnt = 8;
         }
         /* 4 byte response */
         else
         {
            edat32 = (uint32)edat64;
            put_unaligned32(edat32, &(context->esibuf[eadr << 1]));
            cnt = 4;
         }

**Comment by nakarlsson** [comment]
2 minor things
1. Remove ethercattype.h changes from the PR, delete first commit or sqush them?
2. Try to make the uint64 -> uint32 conversion more obvious, I made an inline comment with proposal.

Otherwise it looks OK and works as expected on Little Endian.

File: soem/ethercatmain.c (hunk 4 above)
Comment (Schlumpf) [reply soem/ethercatmain.c]: You are right, will do

**Comment by Schlumpf** [comment]
Done :)

File: soem/ethercattype.h (hunk 3 above)
Comment (Schlumpf) [reply soem/ethercattype.h]: Done

**Comment by nakarlsson** [comment]
thanks, I'll close #174 for now, if more issues are left please write a new issue
